src/System.Private.CoreLib/src/System/Utf8String.Enumeration.cs

```diff
@@ -0,0 +1,289 @@
+
+using System.Buffers;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+
+#pragma warning disable SA1121 // explicitly using type aliases instead of built-in types
+#if BIT64
+using nint = System.Int64;
+using nuint = System.UInt64;
+#else
+using nint = System.Int32;
+using nuint = System.UInt32;
+#endif
+
+namespace System
+{
+    public sealed partial class Utf8String
+    {
+        /*
+         * ENUMERABLE PROPERTY ACCESSORS
+         */
+
+        /// <summary>
+        /// Allows enumeration over the individual UTF-8 bytes of this <see cref="Utf8String"/> instance.
+        /// </summary>
+        public ByteEnumerable Bytes => new ByteEnumerable(this);
+
+        /// <summary>
+        /// Allows enumeration over the UTF-16 code units (see <see cref="char"/>) which would result
+        /// from transcoding this <see cref="Utf8String"/> instance to a UTF-16 <see cref="string"/>.
+        /// </summary>
+        public CharEnumerable Chars => new CharEnumerable(this);
+
+        /// <summary>
+        /// Allows enumeration over the Unicode scalar values (see <see cref="Rune"/>) which are
+        /// encoded by this <see cref="Utf8String"/> instance.
+        /// </summary>
+        public RuneEnumerable Runes => new RuneEnumerable(this);
+
+        /*
+         * ENUMERATORS
+         */
+
+        public readonly struct ByteEnumerable : IEnumerable<byte>
+        {
+            private readonly Utf8String _obj;
+
+            internal ByteEnumerable(Utf8String obj)
+            {
+                _obj = obj;
+            }
+
+            public Enumerator GetEnumerator() => new Enumerator(_obj);
+
+            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+            IEnumerator<byte> IEnumerable<byte>.GetEnumerator() => GetEnumerator();
+
+            public struct Enumerator : IEnumerator<byte>
+            {
+                private readonly Utf8String _obj;
+                private int _curByteIdx;
+
+                internal Enumerator(Utf8String obj)
+                {
+                    _obj = obj;
+                    _curByteIdx = -1;
+                }
+
+                public byte Current
+                {
+                    get
+                    {
+                        byte retVal = default;
+
+                        // Make copies of fields to avoid tearing issues.
+                        // Disallow reading past the end of the object.
+
+                        ReadOnlySpan<byte> byteSpan = _obj.AsBytesSkipNullCheck();
+                        int curByteIdx = _curByteIdx;
+
+                        if ((uint)curByteIdx < (uint)byteSpan.Length)
+                        {
+                            retVal = byteSpan[curByteIdx];
+                        }
+
+                        return retVal;
+                    }
+                }
+
+                public bool MoveNext()
```

Comment:
An idea: store "current" as field and update it here, as all data is local.
Pro: `Current` is just a field access
Con: the struct gets larger (maybe)

(in `Reset` `default` has to be set for the field to macht current implementation of `Current`)

I could imagine that for typical usage of the enumerator with foreach this may have positive effect.

PS: just saw other enumerators use this outlined approach.

Reply:
Honestly might be easier simply to throw NSE from the `Reset` method. It's not strictly required to be implemented, after all.